golem_messages/datastructures/__init__.py

```diff
@@ -0,0 +1,162 @@
+import collections
+import copy
+import enum
+import typing
+
+from golem_messages import exceptions
+
+MessageHeader = collections.namedtuple(
+    "MessageHeader",
+    ["type_", "timestamp", "encrypted"],
+)
+
+
+class SetItemDict(dict):
+    """
+    Mimics dict initialization but always uses __setitem__
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        if args:
+            if isinstance(args[0], dict):
+                for key, value in args[0].items():
+                    self[key] = value
+            else:
+                for key, value in args[0]:
+                    self[key] = value
+        for key in kwargs:
+            self[key] = kwargs[key]
+
+
+class FrozenDict(SetItemDict):
+    """
+    FrozenDict allows only keys provided in ITEMS attribute.
+
+    It also populates any missing keys with defaults from ITEMS.
+    """
+
+    ITEMS = {}
+
+    def _set_defaults(self):
+        for k, v in self.ITEMS.items():
+            if k not in self:
+                self[k] = copy.deepcopy(v)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._set_defaults()
+
+    def __setitem__(self, key, value):
+        if key not in self.ITEMS:
+            raise KeyError("Invalid key: {}".format(key))
+        return super().__setitem__(key, value)
+
+    def __setattr__(self, key, value):
+        raise AttributeError("Read only. Use mapping interface")
+
+
+class ValidatingDict(SetItemDict):
+    """
+    Adds optional validation to the dict interface.
+
+    To add validation to a dictionary key,
+    add a `validate_<key>` method, accepting a value
+    the return value is ignored and the method
+    should raise `exceptions.FieldError`
+    if the value doesn't pass the validation.
+    """
+
+    def __setitem__(self, key, value):
+        validator = getattr(
+            self, 'validate_{}'.format(key), None
+        )
+
+        if callable(validator):
+            validator(value=value)  # noqa pylint:disable=not-callable
+
+        super().__setitem__(key, value)
+
+
+class StringEnum(str, enum.Enum):
+
+    # pylint: disable=no-self-argument
+
+    def _generate_next_value_(name: str, *_):
+        return name
+
+    # pylint: enable=no-self-argument
+
+
+class Container:
+    """Container implementation supporting serialization and deserialization"""
+    # Values in __slots__ are iterables of validators.
+    # Each validator is called with (field_name, value)
+    # See: deserialize_slot()
+    __slots__: typing.Dict[str, typing.callable] = {}
+    # List of required slot names
+    REQUIRED: typing.FozenSet[str] = frozenset()
+    # Default values for slots. See: load_slots
+    DEFAULTS: typing.Dict[str, typing.ANY] = {}
```

Comment:
Gotcha! Values of `DEFAULTS` are not any type, but callables, that return the default value when called without arguments.